### PR TITLE
Create2

### PR DESCRIPTION
**Pull request by marcocastignoli**
image
image

### CODE REVIEW

**Comment by marcocastignoli** [comment]
// TODO
* add tests
* improve UI

**Comment by marcocastignoli** [comment]
commit to add create2 documentation on docs https://github.com/sourcifyeth/docs/commit/50d96860061d929bde75d50f0990cd426684bc8b

File: services/verification/src/services/VerificationService.ts [hunk 1]
```diff
@@ -122,4 +132,64 @@ export class VerificationService implements IVerificationService {
...
+      contract,
+      deployerAddress,
+      salt,
+      constructorArgs,
+      create2Address
+    );
+  };
+
+  recompile = async (contract: CheckedContract): Promise<any> => {
+    // Injection
```
Comment (kuzdogan) [inline services/verification/src/services/VerificationService.ts:165]: comments?

File: services/verification/src/services/VerificationService.ts [hunk 2]
```diff
@@ -122,4 +132,64 @@ export class VerificationService implements IVerificationService {
...
   };
+
+  verifyCreate2 = async (
+    contract: CheckedContract,
+    deployerAddress: string,
+    salt: string,
+    constructorArgs: any,
+    create2Address: string
+  ): Promise<Match> => {
+    // Injection
```
Comment (kuzdogan) [inline services/verification/src/services/VerificationService.ts:143]: comments?

File: ui/src/pages/Lookup/Result.tsx [hunk 3]
```diff
@@ -204,9 +204,68 @@ const InfoText = (
       docs
     </a>{" "}
     for details.
+    {isCreate2Verified && (
+      <p className="mt-3">
+        This contract was verified with the create2 verification, it may not yet
+        be deployed.
```
Comment (kuzdogan) [inline ui/src/pages/Lookup/Result.tsx:210]: Let's add a docs link here

**Review by kuzdogan — CHANGES_REQUESTED** [review]
Added comments

**Comment by kuzdogan** [comment]
Shouldn't there be constructor argument fields?
Screenshot 2022-11-09 at 10 31 20

**Comment by marcocastignoli** [comment]
> Shouldn't there be constructor argument fields?
Are you sure your contract has arguments in the constructor? Can you share the sources?

File: services/core/src/utils/CheckedContract.ts [hunk 4]
```diff
@@ -268,7 +279,7 @@ async function performFetch(
 
   if (res && res.status === 200) {
     const content = await res.text();
-    if (Web3.utils.keccak256(content) !== hash) {
+    if (hash !== null && Web3.utils.keccak256(content) !== hash) {
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply services/core/src/utils/CheckedContract.ts:282]: right

**Comment by kuzdogan** [comment]
> > Shouldn't there be constructor argument fields?
> 
> Are you sure your contract has arguments in the constructor? Can you share the sources?

My bad, doesn't have constructor args

File: services/verification/src/services/VerificationService.ts [hunk 5]
```diff
@@ -122,4 +132,64 @@ export class VerificationService implements IVerificationService {
...
+        fileService: this.fileService,
+        web3timeout: parseInt(process.env.WEB3_TIMEOUT),
+      });
+    }
+
+    return await this.injector.recompile(contract);
+  };
+
+  getBytecode = async (address: string, chainId: string): Promise<any> => {
+    // Injection
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply services/verification/src/services/VerificationService.ts:181]: I don't undestand this

[the remark this answers is not in the document]
Comment (kuzdogan) [reply services/verification/src/services/VerificationService.ts:181]: As in the lines above, I guess these comments (181-183) are forgotten here

File: services/verification/src/services/Injector.ts [hunk 6]
```diff
@@ -557,6 +559,98 @@ export class Injector {
     }
   }
 
+  public async storeMatch(
+    chain: string,
+    addresses: string[],
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply services/verification/src/services/Injector.ts:564]: This is a new function that I made that is used from old functions, I simply generalized some code. Actually by looking at it the `addresses` parameter is completally useless, it is just used for loggin... Can we remove it?

[the remark this answers is not in the document]
Comment (kuzdogan) [reply services/verification/src/services/Injector.ts:564]: Yes sure

**Review by marcocastignoli — COMMENTED** [review]
@kuzdogan can you confirm this?

File: scripts/find_replace.sh [hunk 7]
```diff
@@ -41,6 +42,7 @@ if [ "$CIRCLE_BRANCH" == "master" ]; then
     ESTUARY_PINNING_SECRET=$ESTUARY_PINNING_SECRET_MASTER
     WEB3_STORAGE_PINNING_SECRET=$WEB3_STORAGE_PINNING_SECRET_MASTER
     ETHERSCAN_API_KEY=$ETHERSCAN_API_KEY_MASTER
+    CREATE2_CLIENT_TOKENS=$CREATE2_CLIENT_TOKENS_MASTER
 fi
 
 for VAR_NAME in INFURA_ID ALCHEMY_ID AWS_S3_ACCESS_KEY_ID AWS_S3_SECRET_ACCESS_KEY IPFS_SECRET NPM_TOKEN PUBLIC_IP LOCAL_IP SESSION_SECRET ALCHEMY_ID_OPTIMISM ALCHEMY_ID_ARBITRUM CHAINSAFE_S3_ACCESS_KEY_ID CHAINSAFE_S3_SECRET_ACCESS_KEY ESTUARY_PINNING_SECRET WEB3_STORAGE_PINNING_SECRET ETHERSCAN_API_KEY
```
Comment (kuzdogan) [inline scripts/find_replace.sh:48]: Needs to be added to to-be-replaced variables
```suggestion
for VAR_NAME in INFURA_ID ALCHEMY_ID AWS_S3_ACCESS_KEY_ID AWS_S3_SECRET_ACCESS_KEY IPFS_SECRET NPM_TOKEN PUBLIC_IP LOCAL_IP SESSION_SECRET ALCHEMY_ID_OPTIMISM ALCHEMY_ID_ARBITRUM CHAINSAFE_S3_ACCESS_KEY_ID CHAINSAFE_S3_SECRET_ACCESS_KEY ESTUARY_PINNING_SECRET WEB3_STORAGE_PINNING_SECRET ETHERSCAN_API_KEY CREATE2_CLIENT_TOKENS
```

File: services/verification/src/services/Injector.ts [hunk 8]
```diff
@@ -561,7 +562,6 @@ export class Injector {
 
   public async storeMatch(
     chain: string,
-    addresses: string[],
     contract: CheckedContract,
     compilationResult: RecompilationResult,
     match: Match
```
Comment (kuzdogan) [inline services/verification/src/services/Injector.ts:567]: Not about this PR but it seemed weird to have an `address` in `Match` but not `chain`. WDYT?

File: src/server/controllers/VerificationController.ts [hunk 9]
```diff
@@ -713,6 +851,50 @@ export default class VerificationController
     res.send(getSessionJSON(session));
   };
 
+  private addInputContractEndpoint = async (req: Request, res: Response) => {
```
Comment (kuzdogan) [inline src/server/controllers/VerificationController.ts:854]: What was the purpose of this function? I couldn't find it in the docs either.

File: src/server/controllers/VerificationController.ts [hunk 10]
```diff
@@ -713,6 +851,50 @@ export default class VerificationController
...
+    this.validateRequest(req);
+
+    const address = req.body.address;
+    const chainId = req.body.chainId;
+
+    // TODO: exctract bytecode
```
Comment (kuzdogan) [inline src/server/controllers/VerificationController.ts:860]: TODO left

File: src/server/controllers/VerificationController.ts [hunk 11]
```diff
@@ -713,6 +851,50 @@ export default class VerificationController
...
+
+    if (!metadataIpfsCid) {
+      throw new BadRequestError(
+        "The contract doesn't have a metadata IPFS CID"
+      );
+    }
+
+    const ipfsUrl = `${IPFS_GATEWAY}${metadataIpfsCid}`;
+    const metadataFileName = "metadata.json";
+    const retrievedMetadataText = await performFetch(ipfsUrl);
```
Comment (kuzdogan) [inline src/server/controllers/VerificationController.ts:876]: If we are only using `performFetch` here without hash, why don't we just use a normal `fetch()`?

File: services/core/src/utils/CheckedContract.ts [hunk 12]
```diff
@@ -253,10 +264,10 @@ export class CheckedContract {
  * @param log whether or not to log
  * @returns the fetched file if found; null otherwise
  */
-async function performFetch(
+export async function performFetch(
```
Comment (kuzdogan) [inline services/core/src/utils/CheckedContract.ts:267]: I see the last change resolves my prev. comment of the case `hash === null`. But isn't the whole purpose of the function to do the fetch and compare with the hash? Without the hash it is no different than a standard `fetch()` it seems. See the other comment in `VerificationController`. 

Also currently the function comments do not reflect the functionality.

Is the only reason you added `hash` as optional is to use it in the case at `VerificationController:876`? I feel we should've kept this function as is and done a simple fetch at `VerificationController:876`. Maybe a more descriptive name the a simple `performFetch` would be better, something like `fetchAndCompareHash`.

**Comment by kuzdogan** [comment]
Added a few more comments. The rest look good 👍

File: services/verification/src/services/Injector.ts (hunk 8 above)
Comment (marcocastignoli) [reply services/verification/src/services/Injector.ts:567]: Indeed, maybe we solve this on staging once this PR is merged?

File: src/server/controllers/VerificationController.ts (hunk 9 above)
Comment (marcocastignoli) [reply src/server/controllers/VerificationController.ts:854]: Oh I forgot to add it in the docs, it takes as input an address and a chain of a deployed contract, and imports all of its files in the session

File: services/core/src/utils/CheckedContract.ts (hunk 12 above)
Comment (marcocastignoli) [reply services/core/src/utils/CheckedContract.ts:267]: I think it's a way to always log when fetching something, use the timeout and log errors

Comment (kuzdogan) [reply services/core/src/utils/CheckedContract.ts:267]: Ok, then just updating the comment accordingly should suffice

File: services/verification/src/services/Injector.ts (hunk 8 above)
Comment (marcocastignoli) [reply services/verification/src/services/Injector.ts:567]: Created https://github.com/ethereum/sourcify/issues/834  to keep track of this

**Comment by hoanglamcoin** [comment]
Can you simply and have more pictures for me?  I am very grateful for this and.  I'm pausing this because I don't know what to do.

**Comment by marcocastignoli** [comment]
@hoanglamcoin 
> Can you simply and have more pictures for me? I am very grateful for this and. I'm pausing this because I don't know what to do.

It will soon be available on staging.sourcify.dev - just check in a few days :smile:

**Comment by hoanglamcoin** [comment]
> @hoanglamcoin 
> 
> > Can you simply and have more pictures for me? I am very grateful for this and. I'm pausing this because I don't know what to do.
> 
> 
> 
> It will soon be available on staging.sourcify.dev - just check in a few days :smile: 

what does that mean ?  I don't understand, I have property but I don't know how to verify it.  I used to paste the token and it was there but now it doesn't show anything!  Just because I'm so worried about it.

**Comment by marcocastignoli** [comment]
@hoanglamcoin can you explain me what you are trying to achieve?

you can check the documentation here: https://github.com/sourcifyeth/docs/blob/create2/docs/3.%20Verification/6.create2Verification.md

**Comment by hoanglamcoin** [comment]
Thanks so much!!!!

**Review by kuzdogan — CHANGES_REQUESTED** [review]
Final review, just some leftover comments, and TODOs. Once TODOs are resolved, good to go!

**Review by kuzdogan — APPROVED** [review]

**Pull request merged**